File: T1.SqlSharp/T1.SqlSharp/Expressions/SqlUnpivotClause.cs
using T1.Standard.IO;

namespace T1.SqlSharp.Expressions;

public class SqlUnpivotClause : ISqlExpression
{
    public SqlType SqlType => SqlType.UnpivotClause;
    public TextSpan Span { get; set; } = new();
    public void Accept(SqlVisitor visitor)
    {
        visitor.Visit_UnpivotClause(this);
    }

    public required ISqlExpression NewColumn { get; set; }
    public required ISqlExpression ForSource { get; set; }
    public List<ISqlExpression> InColumns { get; set; } = [];
    public string AliasName { get; set; } = string.Empty; 

    public string ToSql()
    {
        var sql = new IndentStringBuilder();
        sql.Write("UNPIVOT (");
        sql.WriteLine();
        sql.Indent++;
        sql.Write(NewColumn.ToSql());
        sql.WriteLine();
        sql.Write("FOR ");
        sql.Write(ForSource.ToSql());
        sql.WriteLine();
        sql.Write("IN (");
        sql.WriteLine();
        sql.Indent++;
        foreach (var item in InColumns.Select((column, index) => new { column, index }))
        {
            sql.Write(item.column.ToSql());
            if (item.index < InColumns.Count - 1)
            {
                sql.Write(",");
            }
            sql.WriteLine();
        }
        sql.Indent--;
        sql.Write(") ");
        sql.WriteLine($"AS {AliasName}");
        sql.Indent--;
        return sql.ToString();
    }
}

File: VimSharp/chrome-extension-sharp/src/t1-sqlts/Expressions/SqlUnpivotClause.ts
import { SqlType } from './SqlType';
import { TextSpan } from '../StringParser';
import { ISqlExpression, SqlVisitor } from './ISqlExpression';

export class SqlUnpivotClause implements ISqlExpression {
    SqlType: SqlType = SqlType.UnpivotClause;
    Span: TextSpan = new TextSpan();
    NewColumn!: ISqlExpression;
    ForSource!: ISqlExpression;
    InColumns: ISqlExpression[] = [];
    AliasName: string = '';

    Accept(visitor: SqlVisitor): void {
        visitor.Visit_UnpivotClause(this);
    }

    ToSql(): string {
        let sql = 'UNPIVOT (\n';
        sql += this.NewColumn.ToSql() + '\n';
        sql += 'FOR ' + this.ForSource.ToSql() + '\n';
        sql += 'IN (' + this.InColumns.map(x => x.ToSql()).join(', ') + ') ';
        sql += `AS ${this.AliasName}`;
        sql += ')';
        return sql;
    }
} 